var/spack/repos/builtin/packages/llvm/package.py

```diff
@@ -305,45 +335,85 @@ class Llvm(CMakePackage, CudaPackage):
     # Github issue #4986
     patch("llvm_gcc7.patch", when="@4.0.0:4.0.1+lldb %gcc@7.0:")
 
-    # sys/ustat.h has been removed in favour of statfs from glibc-2.28. Use fixed sizes:
-    patch("llvm5-sanitizer-ustat.patch", when="@4:6.0.0+compiler-rt")
-
     # Fix lld templates: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=230463
     patch("llvm4-lld-ELF-Symbols.patch", when="@4+lld%clang@6:")
     patch("llvm5-lld-ELF-Symbols.patch", when="@5+lld%clang@7:")
 
-    # Fix missing std:size_t in 'llvm@4:5' when built with '%clang@7:'
-    patch("xray_buffer_queue-cstddef.patch", when="@4:5+compiler-rt%clang@7:")
+    # fix missing ::size_t in 'llvm@4:5'
+    # see comments in the patch file
+    patch(
+        "xray_buffer_queue-cstddef.patch",
+        # we do not cover compiler-rt=runtime because it is not supported when @:5
+        when="@4:5 compiler-rt=project",
+    )
 
-    # https://github.com/llvm/llvm-project/commit/947f9692440836dcb8d88b74b69dd379d85974ce
-    patch("sanitizer-ipc_perm_mode.patch", when="@5:7+compiler-rt%clang@11:")
-    patch("sanitizer-ipc_perm_mode.patch", when="@5:9+compiler-rt%gcc@9:")
+    # fix building of older versions of llvm with newer versions of glibc
+    for compiler_rt_as in ["project", "runtime"]:
+        with when("compiler-rt={0}".format(compiler_rt_as)):
+            # sys/ustat.h has been removed in favour of statfs from glibc-2.28
+            # see https://reviews.llvm.org/D47281
+            patch(
+                "https://github.com/llvm/llvm-project/commit/383fe5c8668f63ef21c646b43f48da9fa41aa100.patch?full_index=1",
+                sha256="66f01ac1769a6815aba09d6f4347ac1744f77f82ec9578a1158b24daca7a89e6",
+                when="@4:6.0.0",
+            )
+            # fix sanitizer-common build with glibc 2.31
+            # see https://reviews.llvm.org/D70662
+            patch("sanitizer-ipc_perm_mode.patch", when="@5:9")
+    del compiler_rt_as
```

Comment:
Is the result of this that we get the same patches in the same cases? It looks like it but I haven't checked through the llvm commit output to be completely sure.

Reply:
I guess I should give a summary of the patch modifications:
1. `libomp-libflags-as-list.patch` - same conditions, same contents fetched from upstream.
2. `llvm14-hwloc-ompd.patch` - replaced with a "proper" fix from upstream, the old condition was `@14`, the new one is `@14 omp=runtime` (I think I've checked that it's not needed when `omp=project` but will have to double check that.).
3. `constexpr_longdouble.patch` and `constexpr_longdouble_9.0.patch` - same conditions (the old one was `+libcxx`, the new ones are `libcxx=project` and `libcxx=runtime`), same contents fetched from upstream.
4. `llvm5-sanitizer-ustat.patch` - same conditions (the old one was `+compiler-rt`, the new ones are `compiler-rt=project` and `compiler-rt=runtime`), same contents fetched from upstream.
5. `sanitizer-ipc_perm_mode.patch` - replaced with a "proper" fix from upstream but cannot be fetched since the names of the patched files have changed, the old conditions were `@5:7+compiler-rt%clang@11:` and `@5:9+compiler-rt%gcc@9:`, the new ones are `@5:9 compiler-rt=project` and `@5:9 compiler-rt=runtime` (the fixes issue has to do with the version of `glibc` and not the versions of the compiler).
6. `llvm_python_path.patch` - same contents fetched from upstream, the old condition was `@:11`, the new one is `@7:11` (the patched logic was added in `7.0.0`).
7. `xray_buffer_queue-cstddef.patch` - same contents, fixed comments why the patch is needed, no identical commit upstream (nothing to fetch), the old condition was `@4:5+compiler-rt%clang@7:`, the new one is `@4:5 compiler-rt=project` (`compiler-rt=runtime` is not possible when `@:5` and the issue is related to the implementation of the C++ standard library, not the compiler version really).
8. `lldb_external_ncurses-10.patch` - same contents fetched from upstream, the old condition was `@10.0.0:11.0.1+lldb`, the new one is `@10:11+lldb` (other `11.x` versions need it too).
9. `llvm_py37.patch` - same contents fetched from upstream, the old condition was `@4:6 ^python@3.7:`, the new one is `@4:6+python+lldb ^python@3.7:` (the patch is not needed when either `~python` or `~lldb`).

I think that's it.